p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,388 @@
+package holepunch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/holepunch/pb"
+	"github.com/libp2p/go-libp2p/p2p/protocol/identify"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// Protocol is the libp2p protocol for Hole Punching.
+const Protocol protocol.ID = "/libp2p/dcutr"
+
+// StreamTimeout is the timeout for the hole punch protocol stream.
+var StreamTimeout = 1 * time.Minute
```

Comment:
Don't we do this _everywhere_ in libp2p? But I agree with you, config options are better.

Reply:
Unfortunately, this is really handy in the tests.

Reply:
> Don't we do this everywhere in libp2p? But I agree with you, config options are better.

We do, I'd just like to do it less often.

> Unfortunately, this is really handy in the tests.

Can we keep it private? And/or is there an example of a case where an option won't work?

Reply:
> And/or is there an example of a case where an option won't work?

An option would work, but I doubt this is something anyone would like to configure.

> Can we keep it private?

Nope, unless we want to create a separate `internal` package for that. The tests need to use a separate package to avoid an import loop.

Reply:
> An option would work, but I doubt this is something anyone would like to configure.

Yeah... I'm just worried about making it a part of the public API. But it's not too terrible.